### ISSUE TITLE
Possible data loss caused by push operation

### ISSUE THREAD

**Issue by hgjunip** [Azure/azure-mobile-apps#817]
## Describe the bug

We have noticed that sometimes in our app some information goes missing. More specifically the user changes an item so a sync starts for those changes and at that time the user does more changes but when the previously started sync finishes the last changes get overwritten with the previous version. Our users spend a lot of times with intermittent internet connections which might delay responses from the server and make this issue worse. 

## To Reproduce

What we did to reproduce the issue was add a delay on our Create and Replace endpoints. Then we:

1. Modified an item changing its description property to "text A"
2. Started a sync
3. Modified the item again changing its description to "text B"
4. When the sync finished it overwrote the items description to "text A"

## Expected behavior

Items latest changes should not be overwritten since it has pending operations 

## What platforms?

We are using version 6.0.5 in both client and server. The client is a WPF app.

## Additional context

I noticed that when we do a pull we do a check for pending operations [here](https://github.com/Azure/azure-mobile-apps/blob/0bafb6c742602802188cb77f24885c97dde7b370/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/SyncContext.cs#L371C1-L376C22) which prevents the loss of data but when pushing we upsert the result without checking for pending operations [here](https://github.com/Azure/azure-mobile-apps/blob/0bafb6c742602802188cb77f24885c97dde7b370/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/SyncContext.cs#L861C1-L872C14). At this point we haven't confirmed that is this line of code that overwrites the changes but we know that it is the response from the first push of the item to the cloud that triggers the overwrite so seems like the most probable cause. I could try and write a test for it if needed but wanted to know if this sounds like something that could be happening?

**Comment by adrianhall** [comment]
You are correct - we no longer have table locks because everyone complained and it was viewed as the lesser of two evils.  Some things just bite you much later.

Short term, you can put a table lock around your push code so that the database is not updated during a synchronization.

Mid-term, I'll re-institute the table lock that was removed because it caused synchronization to be too slow for everyone.

The correct way of fixing this is in v8, when we switch over to Entity Framework Core for offline.  That will do the proper change tracking of the operations queue.  That will be the longer term plan.

**Comment by hgjunip** [comment]
Could you help me understand the reasoning behind the upsert after a successful push. Is it to update the version? Would doing a pull after a push also work instead of doing the upsert there?

**Comment by adrianhall** [comment]
The push returns the updated value, so we don’t need to do a pull.  If we don’t update at that point, there is an increased chance of unresolvable conflicts.  

I believe we just need to lock the operations queue for that particular ID as the push happens.  I think I can make that happen pretty quickly.

**Comment by hgjunip** [comment]
In our app we already have quite a bit of code to handle conflicts. I noticed that the upsert only happens if the id field is not null on the push result. So yesterday as a work around I tried returning a null id for accepted pushes and since we always pull after a push we can leverage our conflict solving logic that is already in place. The few tests I did seemed to work fine but was wondering if there were any major issues with the approach that I might be missing? 

I worry about how a lock could again affect performance or/and the user experience?
